### PR TITLE
xdsresolver: use the refcounted utility for cluster refcounting

### LINKED ISSUE #9304: xdsresolver: change refcounting for the cluster to use refcounted utility
We have recently introduced a new [refcounted utility](https://github.com/grpc/grpc-go/blob/master/internal/grpcsync/refcounted.go) for refcounting and executing the given function when the refcount goes to zero. 
We can re-use that utility to be used to refcount the clusters in the [xds_resolver.](https://github.com/grpc/grpc-go/blob/1f4c6f36b067b48625280cc7f54cd6c29f058fc5/internal/xds/resolver/xds_resolver.go#L562)

### PR DESCRIPTION
**Pull request by ulascansenturk**
Replaces the hand-rolled `atomic.Int32` in `clusterInfo` with `grpcsync.RefCounted`, as suggested in #9304.

The "decrement, and if the count hit zero run the cleanup" logic was duplicated at three sites (`SelectConfig`'s `OnCommitted`, `configSelector.stop`, and `pruneActiveClustersAndPlugins`). It now lives in a single `onZero` callback registered when the entry is created, which lets `SelectConfig`'s cluster and plugin branches collapse into one path.

Entries now remove themselves from `activeClusters`/`activePlugins` once their last reference is released, so `pruneActiveClustersAndPlugins` is gone.

Two details worth a reviewer's attention:

- **The map removal is scheduled on the serializer**, because the last reference is usually released by an RPC completing on an arbitrary goroutine while the active maps may only be touched from a serializer callback. It is guarded by an identity check so that a pending removal cannot evict a newer entry that has since taken the same key, and acquisition uses `TryIncrement` so a dead entry is replaced rather than revived.

- **For clusters, the removal is queued before `unsubscribe()` is called, and `unsubscribe()` itself stays synchronous.** Queueing first matters because unsubscribing makes the dependency manager push an update onto the same serializer; if the removal were queued after, that update would regenerate a service config still containing the dropped cluster (`TestResolverKeepWatchOpen_ActiveRPCs` catches this). Keeping `unsubscribe()` out of the scheduled callback matters because that callback returns early when a newer entry has taken the key, which would otherwise leak the old subscription.

References for a config selector are now taken as each cluster is recorded, rather than in a batch after all routes are built. This keeps acquisition and release symmetric, so a config selector that fails partway through releases exactly what it acquired.

`TestPruneActiveClusters` is replaced by two tests covering entry reuse/release and the no-revival guard.

Fixes #9304

RELEASE NOTES: none

### CODE REVIEW

File: internal/xds/resolver/xds_resolver.go [hunk 1]
```diff
@@ -425,16 +419,23 @@ func (r *xdsResolver) newConfigSelector() (_ *configSelector, err error) {
 				}
 				return nil, err
 			}
+			ci, ok := cs.plugins[clusterName]
```
Comment (eshitachandwani) [inline internal/xds/resolver/xds_resolver.go:431]: Here , if the plugin/cluster is already present , we are not incrementing the refcount, but we are decrementing the refcount in `cs.stop` unconditionally. I think this will cause discrepencies. We should do a `TryIncrement` even if the plugin is present in the map. Or we can unconditionally call `r.acquireActiveClusterInfo` since that checks the map and increments the refcount. 
WDYT ?

File: internal/xds/resolver/xds_resolver.go [hunk 2]
```diff
@@ -449,16 +450,23 @@ func (r *xdsResolver) newConfigSelector() (_ *configSelector, err error) {
 					}
 					return nil, err
 				}
+				ci, ok := cs.clusters[clusterName]
```
Comment (eshitachandwani) [inline internal/xds/resolver/xds_resolver.go:464]: Same as plugin

File: internal/xds/resolver/xds_resolver.go [hunk 3]
```diff
@@ -502,66 +500,92 @@ func (r *xdsResolver) newConfigSelector() (_ *configSelector, err error) {
...
+		r.scheduleActiveEntryRemoval(r.activeClusters, key, ci, nil)
+		// Unsubscribing decrements the reference count in the dependency
+		// manager; once that count reaches zero, the underlying CDS watch is
+		// terminated.
+		//
+		// This runs synchronously rather than inside the scheduled removal: that
+		// removal returns early if a newer entry has taken this key, and the
+		// newer entry holds its own subscription, so deferring the unsubscribe
+		// would leak this one.
+		unsubscribe()
```
Comment (eshitachandwani) [inline internal/xds/resolver/xds_resolver.go:565]: Running unsubscribe here would mean that it can run before actually removing the entry from the activeClusters map , because TrySchedule will just schedule and return, which might cause wrong update being sent.
We might want to use `ScheduleAndWait`. WDYT ?

**Comment by eshitachandwani** [comment]
Hey @ulascansenturk, could you please reply to the comments once you've addressed them? Also, when updating the PR, please push your new changes as separate commits rather than force-pushing to squash them. Keeping the commit history intact makes it much easier for us to review what has changed. Thanks!

File: internal/xds/resolver/xds_resolver.go (hunk 1 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver.go]: I checked this and I believe the counts do balance, so I have left the logic as is — but the fact that it reads as unbalanced is fair, and I have added a comment at both sites to make the invariant explicit.

`cs.stop()` releases these references by ranging over `cs.clusters` / `cs.plugins`, which are maps keyed by cluster name, so it decrements **once per distinct cluster**, not once per route that named it. The `if !ok` guard takes exactly one reference per distinct cluster, so acquire and release match.

Calling `acquireActiveClusterInfo` unconditionally would increment once per route while `stop()` still decrements once per map entry, so a cluster named by two routes would keep a surplus reference and never be released. I tried it to confirm rather than reason about it: with that change, `TestResolverClusterSharedByMultipleRoutes` (added in this PR — two routes pointing at the same cluster) times out with `cluster-A` still present in the service config and its CDS watch open. It passes as written.

File: internal/xds/resolver/xds_resolver.go (hunk 2 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver.go]: Same reasoning as the plugin case above — `cs.stop()` ranges over the `cs.clusters` map, so it decrements once per distinct cluster and matches the single reference taken here. I have added the same clarifying comment at this site, and `TestResolverClusterSharedByMultipleRoutes` covers the shared-cluster case for this path.

File: internal/xds/resolver/xds_resolver.go (hunk 1 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver.go:431]: The counts do balance here. `cs.stop()` ranges over `cs.plugins`, which is keyed by cluster name, so it decrements once per distinct plugin however many routes named it. The `if !ok` guard takes exactly one ref per distinct plugin, so the two match.

Calling `acquireActiveClusterInfo` unconditionally would increment per route while `stop()` still decrements once per entry, so anything named by two routes would never be released. I tried it, and `TestResolverClusterSharedByMultipleRoutes` (added here) times out with cluster-A still in the service config.

File: internal/xds/resolver/xds_resolver.go (hunk 2 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver.go:464]: Same as above. `cs.stop()` ranges over `cs.clusters`, so it decrements once per distinct cluster and matches the single ref taken here. `TestResolverClusterSharedByMultipleRoutes` covers this path too.

File: internal/xds/resolver/xds_resolver.go (hunk 3 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver.go:565]: Good catch that `unsubscribe()` runs before the entry is actually deleted. It stays safe because the update it provokes goes through the same serializer: the dependency manager calls `r.Update()`, which queues behind the removal already scheduled, so the entry is gone by the time the config is regenerated. Swapping the two makes `TestResolverKeepWatchOpen_ActiveRPCs` fail.

`ScheduleAndWait` deadlocks though. `configSelector.stop()` drops the last ref from inside a serializer callback, so waiting there blocks the serializer against itself. I tried it and the package times out, with the stack showing `Decrement` reaching `ScheduleAndWait` from within `CallbackSerializer.run`.

File: internal/xds/resolver/xds_resolver.go (hunk 1 above)
Comment (eshitachandwani) [reply internal/xds/resolver/xds_resolver.go:431]: Ohh okay! I misunderstood that cs.plugin is a shared state.

File: internal/xds/resolver/serviceconfig.go [hunk 4]
```diff
@@ -109,6 +109,10 @@ type virtualHost struct {
 type routeCluster struct {
 	name        string                       // Name of the cluster.
 	interceptor httpfilter.ClientInterceptor // HTTP filters to run for RPCs matching this route.
+	// info is the resolver-wide entry for this cluster, shared by every route
+	// that references it. An RPC routed here holds a reference on it until the
+	// RPC is committed.
+	info *grpcsync.RefCounted[*clusterInfo]
```
Comment (eshitachandwani) [inline internal/xds/resolver/serviceconfig.go]: Nit: we can name it `clusterInfo` to be a little more explicit.

File: internal/xds/resolver/xds_resolver_test.go [hunk 5]
```diff
@@ -1385,6 +1385,95 @@ func (s) TestResolverKeepWatchOpen_ActiveRPCs(t *testing.T) {
 // TestResolver_XDSConfigInRPCContext verifies that the xDS resolver's config
```
Comment (eshitachandwani) [inline internal/xds/resolver/xds_resolver_test.go]: This the comment has been seperated from the test , please fix.

File: internal/xds/resolver/xds_resolver_test.go [hunk 6]
```diff
@@ -1385,6 +1385,95 @@ func (s) TestResolverKeepWatchOpen_ActiveRPCs(t *testing.T) {
...
 // selector places the complete XDSConfig into the RPC context during config
 // selection, making it available to HTTP filters.
+// TestResolverClusterSharedByMultipleRoutes verifies the reference accounting
+// for a cluster that more than one route points at. A config selector takes a
+// single reference per distinct cluster, however many routes name it, and
+// releases exactly that one reference when it is stopped.
+//
+// If a reference were taken per route instead, the counts would not balance:
```
Comment (eshitachandwani) [inline internal/xds/resolver/xds_resolver_test.go]: I think we can remove this paragraph. We is being done and tested should be enough.

File: internal/xds/resolver/serviceconfig_test.go [hunk 7]
```diff
@@ -44,50 +46,113 @@ func Test(t *testing.T) {
...
+// runOnSerializer runs f in the context of a serializer callback, which is the
+// only place the resolver's active cluster and plugin maps may be touched, and
+// blocks until it has run. Any callback queued by an earlier call to this
+// helper, including a removal queued when a reference count reached zero, is
+// guaranteed to have run by the time f is invoked.
+func runOnSerializer(ctx context.Context, t *testing.T, r *xdsResolver, f func()) {
+	t.Helper()
+
+	done := make(chan struct{})
+	r.serializer.TrySchedule(func(context.Context) {
```
Comment (eshitachandwani) [inline internal/xds/resolver/serviceconfig_test.go]: We have a scheduleAndWait function which is designed to do exactly this, schedule the function and wait for it to be executed. Can we use that instead?

File: internal/xds/resolver/serviceconfig_test.go [hunk 8]
```diff
@@ -44,50 +46,113 @@ func Test(t *testing.T) {
...
 	}
-	wantActivePlugins := map[string]int32{
-		"one": 1,
-		"two": 2,
+}
+
+// TestActivePluginRefCounting verifies that repeated acquisitions of a cluster
+// specifier plugin share a single entry, and that the entry is removed from
+// activePlugins only once the last reference is released.
+func (s) TestActivePluginRefCounting(t *testing.T) {
```
Comment (eshitachandwani) [inline internal/xds/resolver/serviceconfig_test.go]: THis can easily be converted to an E2E test where we send RPCs on a plugin and make sure it is not removed from the service config when the RPCs are in flight. ANd is removed once the RPCs are done and it is no longer part of the new XDSConfig.

File: internal/xds/resolver/serviceconfig_test.go [hunk 9]
```diff
@@ -44,50 +46,113 @@ func Test(t *testing.T) {
...
+			t.Errorf("activePlugins[%q] still present after the last reference was released", key)
+		}
+	})
+}
+
+// TestActivePluginNotRevivedAfterRelease verifies that an entry whose reference
+// count has already dropped to zero is replaced by a fresh entry rather than
+// resurrected, and that the pending removal of the dead entry does not delete
+// its replacement.
+func (s) TestActivePluginNotRevivedAfterRelease(t *testing.T) {
```
Comment (eshitachandwani) [inline internal/xds/resolver/serviceconfig_test.go]: Can we also try to turn this into an e2e test? We might not be able to test the exact same thing but we should be able to test the concurrency , something like if last RPC ref is committed in a go routine and we send the xdsConfig update at the same time, whichever happens first should be handled correctly. WDYT ?

File: internal/xds/resolver/xds_resolver_test.go (hunk 5 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver_test.go]: Fixed. The doc comment for `TestResolver_XDSConfigInRPCContext` had been left above my test, so both were orphaned. Moved it back onto its own function.

File: internal/xds/resolver/xds_resolver_test.go (hunk 6 above)
Comment (ulascansenturk) [reply internal/xds/resolver/xds_resolver_test.go]: Removed.

File: internal/xds/resolver/serviceconfig_test.go (hunk 8 above)
Comment (ulascansenturk) [reply internal/xds/resolver/serviceconfig_test.go]: Dropped this test entirely. `TestXDSResolverDelayedOnCommittedCSP` already covers that scenario end to end: it holds an RPC on cspA, switches the route to cspB, checks both appear in the service config, then commits the RPC and checks cspA is gone. Adding another e2e test for it would just duplicate that one.

File: internal/xds/resolver/serviceconfig_test.go (hunk 9 above)
Comment (ulascansenturk) [reply internal/xds/resolver/serviceconfig_test.go]: Done, added `TestResolverClusterSpecifierPluginRefCountRace` in `cluster_specifier_plugin_test.go` and removed the unit test.

It holds an RPC on cspA, moves the route to cspB so cspA is referenced only by that RPC, then commits the RPC in a goroutine while pushing an update that names cspA again. Both orderings have to converge on cspA being in the service config and usable for new RPCs: if the commit wins, the entry is torn down and `TryIncrement` fails so a fresh one replaces it, and if the update wins the existing entry is reused. Passes with `-race` at `-count=10`.

I also added a `waitForServiceConfig` helper, since the resolver publishes intermediate configs here and `verifyUpdateFromResolver` only looks at the next update.

Comment (ulascansenturk) [reply internal/xds/resolver/serviceconfig_test.go]: Follow up: the first version of this test was flaky and CI caught it. It pushed the cspB update and the cspA update back to back without waiting in between, so on a loaded machine the client could be scheduled only after both snapshots had landed. The route config it then saw was identical to the original cspA one, so the resolver had nothing new to publish and the test timed out. Reproducible locally with `-cpu 1`.

It now waits for the intermediate config naming both plugins before starting the race, which confirms the client processed cspB first. Passing at `-count=20` with `-cpu 1`, and with `-race` at `-cpu` 1, 2 and 4.

**Comment by ulascansenturk** [comment]
Heads up on the red `upload` check, it is not coming from this PR. The coverage workflow pins `go-version: stable`, which now resolves to go1.27.0, while the test matrix pins 1.25 and 1.26. Go 1.27 changed `compress/flate` output, so the gzip sizes hardcoded in the `stats/opentelemetry` tests no longer match. `e2e_test.go:448` expects 57 and gets 54. Checking it directly, gzip of 10000 zero bytes goes from 48 bytes on 1.25 and 1.26 to 45 bytes on 1.27, the same 3 byte delta.

The identical failure is on #9343 and a couple of other open PRs, and master's own codecov runs were green through the 19th, which lines up with when `stable` flipped to 1.27. Every test job on this PR is passing, including the race one.

I could not find an existing issue for it. I am happy to open one and fix it if nobody is already on it. It looks like the hardcoded sizes in the otel and csm tests want computing at runtime rather than asserting a literal, otherwise it breaks again on the next flate change.

**Comment by eshitachandwani** [comment]
> Heads up on the red `upload` check, it is not coming from this PR. The coverage workflow pins `go-version: stable`, which now resolves to go1.27.0, while the test matrix pins 1.25 and 1.26. Go 1.27 changed `compress/flate` output, so the gzip sizes hardcoded in the `stats/opentelemetry` tests no longer match. `e2e_test.go:448` expects 57 and gets 54. Checking it directly, gzip of 10000 zero bytes goes from 48 bytes on 1.25 and 1.26 to 45 bytes on 1.27, the same 3 byte delta.
> 
> The identical failure is on #9343 and a couple of other open PRs, and master's own codecov runs were green through the 19th, which lines up with when `stable` flipped to 1.27. Every test job on this PR is passing, including the race one.
> 
> I could not find an existing issue for it. I am happy to open one and fix it if nobody is already on it. It looks like the hardcoded sizes in the otel and csm tests want computing at runtime rather than asserting a literal, otherwise it breaks again on the next flate change.

Hey , thank you for investigating this, we are already in discussion about the best way to fix this.

**Comment by eshitachandwani** [comment]
Hey @ulascansenturk  , can you rebase it on master to get the codecov job passing ?

**Comment by ulascansenturk** [comment]
Done, merged master in rather than rebasing so the existing commits stay intact, as you asked earlier. That picks up #9346, and the otel and csm packages now pass locally under the same `-coverprofile -coverpkg=./...` the codecov job uses.

Also reran the full `./internal/xds/... ./xds/... ./test/xds/...` with `-race` after the merge and it is green.

File: internal/xds/resolver/xds_resolver.go [hunk 10]
```diff
@@ -506,66 +508,97 @@ func (r *xdsResolver) newConfigSelector() (_ *configSelector, err error) {
...
+			return
+		}
+		delete(m, key)
+		if onRemoved != nil {
+			onRemoved()
+		}
+	})
+}
+
 type clusterInfo struct {
```
Comment (eshitachandwani) [inline internal/xds/resolver/xds_resolver.go:598]: Now that this only has one field, can we remove the struct and just use `xdsChildConfig` instead ?

File: internal/xds/resolver/serviceconfig.go (hunk 4 above)
Comment (eshitachandwani) [inline internal/xds/resolver/serviceconfig.go:115]: Why do we need to store it here ? We can retrieve it from the configSelector struct itself , similar to what was being done earlier. Storing it in different places increases risk and overhead of making sure they are all synced.
